### PR TITLE
Dcutr

### PR DESCRIPTION
**Pull request by diegomrsantos**
This PR partially implements https://github.com/libp2p/specs/blob/master/relay/DCUtR.md. Currently, only the TCP protocol is supported. Additionally, the unilateral connection upgrade is handled by the Hole Punching service instead.

### CODE REVIEW

File: libp2p/dial.nim [hunk 1]
```diff
@@ -22,6 +22,7 @@ export results
 
 type
   Dial* = ref object of RootObj
+    isSimultaneousConnServer*: bool
```
Comment (Menduist) [inline libp2p/dial.nim:25]: This is not good, there may be multiple dials in parallel

File: libp2p/protocols/connectivity/dcutr/client.nim [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import ../../protocol,
+       ../../../stream/connection,
+       ../../../switch
+
+import stew/results
+import chronos, chronicles
+
+type
+  DcutrClient* = ref object of RootObj
+    rttStart: Opt[Moment]
```
Comment (Menduist) [inline libp2p/protocols/connectivity/dcutr/client.nim:25]: Why is this here? This prevents parallel use of the system, and doesn't seem to be used outside of startSync

File: libp2p/protocols/connectivity/dcutr/client.nim [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    rttEnd: Opt[Moment]
+
+logScope:
+  topics = "libp2p dcutrclient"
+
+proc new*(T: typedesc[DcutrClient]): T =
+  return T(rttStart: Opt.none(Moment), rttEnd: Opt.none(Moment))
+
+proc sendSyncMsg(stream: Connection, addrs: seq[MultiAddress]) {.async.} =
```
Comment (Menduist) [inline libp2p/protocols/connectivity/dcutr/client.nim:34]: Seems weird to have this proc here and `sendConnectMsg` in another file

File: libp2p/protocols/connectivity/dcutr/core.nim [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    msgTypeOrd: uint32
+    dcutrMsg: DcutrMsg
+  var pb = initProtoBuffer(buf)
+  var r1 = pb.getField(1, msgTypeOrd)
+  let r2 = pb.getRepeatedField(2, dcutrMsg.addrs)
+  if r1.isErr or r2.isErr or not checkedEnumAssign(dcutrMsg.msgType, msgTypeOrd):
+    raise newException(DcutrError, "Received malformed message")
+  return dcutrMsg
+
+proc sendConnectMsg*(conn: Connection, addrs: seq[MultiAddress]) {.async.} =
```
Comment (Menduist) [inline libp2p/protocols/connectivity/dcutr/core.nim:58]: Maybe instead of this you could have
```nim
proc send*(conn: Connection, msg: DcutrMsg) {.async.} =
  xxx

# Used like:
await conn.send(DcturMsg(msgType: MsgType.Connect, addrs: xxx))
await conn.send(DcturMsg(msgType: MsgType.Sync, addrs: xxx))
```
Less boilerplate

File: libp2p/protocols/connectivity/dcutr/client.nim [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    debug "Dcutr initiator has sent a Connect message."
+    self.rttStart = Opt.some(Moment.now())
+    let connectAnswer = DcutrMsg.decode(await stream.readLp(1024))
+    self.rttEnd = Opt.some(Moment.now())
+    debug "Dcutr initiator has received a Connect message back.", connectAnswer
+    let halfRtt = (self.rttEnd.get() - self.rttStart.get()) div 2
+    await sendSyncMsg(stream, addrs)
+    debug "Dcutr initiator has sent a Sync message."
+    await sleepAsync(halfRtt)
+    await switch.connect(remotePeerId, connectAnswer.addrs, true, false, Direction.In)
```
Comment (Menduist) [inline libp2p/protocols/connectivity/dcutr/client.nim:55]: Just for clarity, could you name the boolean flags here, this is not very clear without reading the connect signature

File: libp2p/protocols/connectivity/dcutr/client.nim [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+else:
+  {.push raises: [].}
+
+import core
+import ../../protocol,
+       ../../../stream/connection,
+       ../../../switch
+
+import stew/results
+import chronos, chronicles
```
[the remark this answers is not in the document]
Comment (Menduist) [reply libp2p/protocols/connectivity/dcutr/client.nim:21]: Also applies to other files

File: libp2p/stream/lpstream.nim [hunk 7]
```diff
@@ -80,6 +80,11 @@ type
     opened*: uint64
     closed*: uint64
 
+proc `$`*(dir: Direction): string =
```
Comment (Menduist) [inline libp2p/stream/lpstream.nim:83]: Why is this necessary?

File: libp2p/upgrademngrs/muxedupgrade.nim [hunk 8]
```diff
@@ -62,14 +62,14 @@ proc mux*(
...
-proc upgrade(
+method upgrade*(
   self: MuxedUpgrade,
   conn: Connection,
   direction: Direction,
   peerId: Opt[PeerId]): Future[Muxer] {.async.} =
   trace "Upgrading connection", conn, direction
 
-  let sconn = await self.secure(conn, direction, peerId) # secure the connection
+  let sconn = await self.secure(conn, direction, if direction == In: Opt.none(PeerId) else: peerId) # secure the connection
```
Comment (Menduist) [inline libp2p/upgrademngrs/muxedupgrade.nim:72]: Why is this necessary?

File: libp2p/protocols/connectivity/dcutr/server.nim [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+proc new*(T: typedesc[Dcutr], switch: Switch): T =
+
+  proc handleStream(stream: Connection, proto: string) {.async, gcsafe.} =
+    try:
+      let connectMsg = DcutrMsg.decode(await stream.readLp(1024))
+      debug "Dcutr receiver received a Connect message.", connectMsg
+      var ourAddrs = switch.peerStore.getMostObservedProtosAndPorts() # likely empty when the peer is reachable
+      if ourAddrs.len == 0:
+        # this list should be the same as the peer's public addrs when it is reachable
+        ourAddrs =  switch.peerInfo.addrs.mapIt(switch.peerStore.guessDialableAddr(it))
```
Comment (Menduist) [inline libp2p/protocols/connectivity/dcutr/server.nim:41]: Can we get in this situation when we are not reachable? If so, how does it handle circuit relay addresses & co?

Comment (Menduist) [reply libp2p/protocols/connectivity/dcutr/server.nim:41]: Instinctively it seems like this should be `switch.peerInfo.listenAddrs` instead of addrs

File: libp2p/protocols/connectivity/dcutr/server.nim [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+      await sendConnectMsg(stream, ourAddrs)
+      debug "Dcutr receiver has sent a Connect message back."
+      let syncMsg = DcutrMsg.decode(await stream.readLp(1024))
+      debug "Dcutr receiver has received a Sync message.", syncMsg
+      await switch.connect(stream.peerId, connectMsg.addrs, true, false)
+      debug "Dcutr receiver has directly connected to the remote peer."
+    except CatchableError as err:
+      error "Unexpected error in dcutr handler", msg = err.msg
+      raise newException(DcutrError, "Unexpected error when trying a direct conn", err)
```
Comment (Menduist) [inline libp2p/protocols/connectivity/dcutr/server.nim:50]: finally: close stream

Comment (Menduist) [reply libp2p/protocols/connectivity/dcutr/server.nim:50]: And same log level things as for the client

File: tests/testdcutr.nim [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    try:
+      # we can't hole punch when both peers are in the same machine. This means that the simultaneous dialings will result
+      # in two connections attemps, instead of one. This dial is going to fail because the dcutr client is acting as the
+      # tcp simultaneous incoming upgrader in the dialer which works only in the simultaneous open case.
+      await DcutrClient.new().startSync(behindNATSwitch, publicSwitch.peerInfo.peerId, behindNATSwitch.peerInfo.addrs)
+      .wait(300.millis)
+    except CatchableError as exc:
+      discard
+
+    await sleepAsync(200.millis) # wait for the dcutr server to finish
```
Comment (Menduist) [inline tests/testdcutr.nim:68]: Why is this necessary?

File: libp2p/protocols/connectivity/dcutr/client.nim [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    self.rttEnd = Opt.some(Moment.now())
+    debug "Dcutr initiator has received a Connect message back.", connectAnswer
+    let halfRtt = (self.rttEnd.get() - self.rttStart.get()) div 2
+    await sendSyncMsg(stream, addrs)
+    debug "Dcutr initiator has sent a Sync message."
+    await sleepAsync(halfRtt)
+    await switch.connect(remotePeerId, connectAnswer.addrs, true, false, Direction.In)
+    debug "Dcutr initiator has directly connected to the remote peer."
+  except CatchableError as err:
+    error "Unexpected error when trying direct conn", err = err.msg
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply libp2p/protocols/connectivity/dcutr/client.nim:58]: The purpose of the debug is to help when debugging issues and it was very helpful in this task. I find trace of very limited help given the huge amount of information produced.

[the remark this answers is not in the document]
Comment (Menduist) [reply libp2p/protocols/connectivity/dcutr/client.nim:58]: With this PR: https://github.com/status-im/nim-chronicles/pull/124 it will be possible to do:
`chronicles_enabled_topics="*:INFO dctur*:TRACE"`

File: libp2p/protocols/connectivity/dcutr/client.nim (hunk 2 above)
Comment (diegomrsantos) [reply libp2p/protocols/connectivity/dcutr/client.nim:25]: True, I think it is from an early version. Removed.

File: libp2p/stream/lpstream.nim (hunk 7 above)
Comment (diegomrsantos) [reply libp2p/stream/lpstream.nim:83]: To do this in the dialer `doAssert not isNil(upgradedConn), "connection died after upgrade " & $upgradeDir`

File: libp2p/upgrademngrs/muxedupgrade.nim (hunk 8 above)
Comment (diegomrsantos) [reply libp2p/upgrademngrs/muxedupgrade.nim:72]: `upgradeIncoming` did this `self.upgrade(conn, In, Opt.none(PeerId))`

File: libp2p/stream/lpstream.nim (hunk 7 above)
Comment (Menduist) [reply libp2p/stream/lpstream.nim:83]: Sounds like a bug / missing import / whatever, `$` should be autogenerated:
```nim
type Xxx {.pure.} = enum
  A, B, C

#works
echo $Xxx.A
```

File: libp2p/upgrademngrs/muxedupgrade.nim (hunk 8 above)
Comment (Menduist) [reply libp2p/upgrademngrs/muxedupgrade.nim:72]: I think we can just forward the PeerId, it will be `Some` during HP, and will actually improve security by checking the peer identity earlier (are you checking peer identity btw currently?)

File: libp2p/stream/lpstream.nim (hunk 7 above)
Comment (diegomrsantos) [reply libp2p/stream/lpstream.nim:83]: I had an issue with that, but now I can't reproduce it. Removed.

File: libp2p/upgrademngrs/muxedupgrade.nim (hunk 8 above)
Comment (diegomrsantos) [reply libp2p/upgrademngrs/muxedupgrade.nim:72]: AFAIK I'm not checking it. Could you please elaborate more on that?

File: libp2p/protocols/connectivity/dcutr/server.nim (hunk 9 above)
Comment (diegomrsantos) [reply libp2p/protocols/connectivity/dcutr/server.nim:41]: > Can we get in this situation when we are not reachable?
Yes, we'll get here when there aren't enough observations (currently at least 3) of the same IP/Port.

Comment (diegomrsantos) [reply libp2p/protocols/connectivity/dcutr/server.nim:41]: > how does it handle circuit relay addresses & co?

I handled this in the HP PR, but should have done it here. Does it make sense?
```
      var dialableAddrs = connectMsg.addrs
      dialableAddrs.keepItIf(not it.contains(multiCodec("p2p-circuit")).valueOr(true))
      if dialableAddrs.len == 0:
        debug "Dcutr receiver has no dialable addresses to connect to."
        return
```

Comment (diegomrsantos) [reply libp2p/protocols/connectivity/dcutr/server.nim:41]: Now I realized you were asking about sending relay addrs. You're right, it should be `listenAddrs`.

File: libp2p/protocols/connectivity/dcutr/server.nim (hunk 10 above)
Comment (diegomrsantos) [reply libp2p/protocols/connectivity/dcutr/server.nim:50]: isn't the stream closed by the code that called `handleStream`?

File: tests/testdcutr.nim [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+    await publicSwitch.connect(behindNATSwitch.peerInfo.peerId, behindNATSwitch.peerInfo.addrs)
+
+    for t in behindNATSwitch.transports:
+      t.networkReachability = NetworkReachability.NotReachable
+
+    for t in publicSwitch.transports:
+      t.networkReachability = NetworkReachability.NotReachable
+
+    try:
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply tests/testdcutr.nim:59]: I wrote I expect it to fail, but it is more like "This is likely going to fail locally, but for now it is fine if in the end a new connection is created." The test should still pass if it doesn't fail.

File: libp2p/protocols/connectivity/dcutr/client.nim (hunk 6 above)
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply libp2p/protocols/connectivity/dcutr/client.nim]: done

File: libp2p/protocols/connectivity/dcutr/server.nim (hunk 10 above)
Comment (Menduist) [reply libp2p/protocols/connectivity/dcutr/server.nim:50]: Only in case of failures, I think

Comment (Menduist) [reply libp2p/protocols/connectivity/dcutr/server.nim:50]: Oh you are right, nvm

File: libp2p/dialer.nim [hunk 14]
```diff
@@ -71,27 +71,29 @@ proc dialAndUpgrade(
...
+            if upgradeDir == Direction.Out:
+              libp2p_failed_upgrades_outgoing.inc()
+            else:
+              libp2p_failed_upgrades_incoming.inc()
 
           # Try other address
           return nil
 
-      doAssert not isNil(mux), "connection died after upgradeOutgoing"
+      doAssert not isNil(mux), "connection died after upgrade " & $upgradeDir
```
Comment (Menduist) [inline libp2p/dialer.nim:96]: This line is redundant with line 80?

File: libp2p/protocols/connectivity/dcutr/client.nim (hunk 3 above)
Comment (Menduist) [reply libp2p/protocols/connectivity/dcutr/client.nim:37]: And so? Still weird

File: libp2p/upgrademngrs/muxedupgrade.nim (hunk 8 above)
Comment (Menduist) [reply libp2p/upgrademngrs/muxedupgrade.nim]: I meant checking that the connection established during HP is indeed with the correct peer (by checking the peerid)

Comment (diegomrsantos) [reply libp2p/upgrademngrs/muxedupgrade.nim]: In this PR we're always giving the remote peer id to the upgrade process.

File: libp2p/protocols/connectivity/dcutr/core.nim (hunk 4 above)
Comment (diegomrsantos) [reply libp2p/protocols/connectivity/dcutr/core.nim]: We agreed to go with:

```nim
proc send*(conn: Connection, msgType: MsgType, addrs: seq[MultiAddress]) {.async.} =
  let pb = DcutrMsg(msgType: msgType, addrs: addrs).encode()
  await conn.writeLp(pb.buffer)
```

File: tests/testdcutr.nim [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+
+import ../libp2p/protocols/connectivity/dcutr/core as dcore
+import ../libp2p/protocols/connectivity/dcutr/[client, server]
+from ../libp2p/protocols/connectivity/autonat/core import NetworkReachability
+import ../libp2p/builders
+import ../libp2p/utils/future
+import ./helpers
+
+type
+  SwitchStub* = ref object of Switch
```
Comment (diegomrsantos) [inline tests/testdcutr.nim:21]: This will be moved to its own file in the HP PR.

File: libp2p/dialer.nim [hunk 16]
```diff
@@ -71,27 +71,28 @@ proc dialAndUpgrade(
...
-      dialed.transportDir = Direction.Out
-
       libp2p_successful_dials.inc()
 
       let mux =
         try:
-          await transport.upgradeOutgoing(dialed, peerId)
+          dialed.transportDir = upgradeDir
+          let upgradedConn = await transport.upgrade(dialed, upgradeDir, peerId)
+          upgradedConn
```
Comment (Menduist) [inline libp2p/dialer.nim:80]:
```suggestion
          await transport.upgrade(dialed, upgradeDir, peerId)
```

File: libp2p/protocols/connectivity/dcutr/client.nim [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+  logScope:
+    peerId = switch.peerInfo.peerId
+
+  var
+    peerDialableAddrs: seq[MultiAddress]
+    stream: Connection
+  try:
+    var ourDialableAddrs = getTCPAddrs(addrs)
+    if ourDialableAddrs.len == 0:
+      debug "Dcutr initiator has no supported dialable addresses. Aborting Dcutr."
```
Comment (Menduist) [inline libp2p/protocols/connectivity/dcutr/client.nim:47]:
```suggestion
      debug "Dcutr initiator has no supported dialable addresses. Aborting Dcutr.", addrs
```

File: libp2p/protocols/connectivity/dcutr/client.nim [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+
+    stream = await switch.dial(remotePeerId, DcutrCodec)
+    await stream.send(MsgType.Connect, addrs)
+    debug "Dcutr initiator has sent a Connect message."
+    let rttStart = Moment.now()
+    let connectAnswer = DcutrMsg.decode(await stream.readLp(1024))
+
+    peerDialableAddrs = getTCPAddrs(connectAnswer.addrs)
+    if peerDialableAddrs.len == 0:
+      debug "DDcutr receiver has no supported dialable addresses to connect to. Aborting Dcutr."
```
Comment (Menduist) [inline libp2p/protocols/connectivity/dcutr/client.nim:58]:
```suggestion
      debug "DDcutr receiver has no supported dialable addresses to connect to. Aborting Dcutr.", addrs=connectAnswer.adds
```

File: libp2p/protocols/connectivity/dcutr/client.nim [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+      return
+
+    let rttEnd = Moment.now()
+    debug "Dcutr initiator has received a Connect message back.", connectAnswer
+    let halfRtt = (rttEnd - rttStart) div 2'i64
+    echo halfRtt.type
```
Comment (Menduist) [inline libp2p/protocols/connectivity/dcutr/client.nim:64]:
```suggestion
```

File: libp2p/protocols/connectivity/dcutr/client.nim [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+
+    if peerDialableAddrs.len > self.maxDialableAddrs:
+        peerDialableAddrs = peerDialableAddrs[0..<self.maxDialableAddrs]
+    var futs = peerDialableAddrs.mapIt(switch.connect(stream.peerId, @[it], forceDial = true, reuseConnection = false, upgradeDir = Direction.In))
+    let fut = await anyCompleted(futs).wait(self.connectTimeout)
+    await fut
+    if fut.completed():
+      debug "Dcutr initiator has directly connected to the remote peer."
+    else:
+      debug "Dcutr initiator could not connect to the remote peer.", msg = fut.error.msg
```
Comment (Menduist) [inline libp2p/protocols/connectivity/dcutr/client.nim:77]:
```suggestion
    try:
      discard await anyCompleted(futs).wait(self.connectTimeout)
      debug "Dcutr initiator has directly connected to the remote peer."
    except CatchableError as err:
      debug "Dcutr initiator could not connect to the remote peer.", msg = err.msg
      raise err
    finally:
      for fut in futs: fut.cancel()
```

File: libp2p/protocols/connectivity/dcutr/core.nim [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+  return dcutrMsg
+
+proc send*(conn: Connection, msgType: MsgType, addrs: seq[MultiAddress]) {.async.} =
+  let pb = DcutrMsg(msgType: msgType, addrs: addrs).encode()
+  await conn.writeLp(pb.buffer)
+
+proc getTCPAddrs*(addrs: seq[MultiAddress]): seq[MultiAddress] =
+  var tcpAddrs = addrs
+  tcpAddrs.keepItIf(TCP.matchPartial(it))
+  return tcpAddrs
```
Comment (Menduist) [inline libp2p/protocols/connectivity/dcutr/core.nim:65]:
```suggestion
proc getTCPAddrs*(addrs: seq[MultiAddress]): seq[MultiAddress] =
  addrs.filterIt(TCP.matchPartial(it))
```
And maybe rename to something more generic, like getHolePunchableAddrs? Idk

File: libp2p/protocols/connectivity/dcutr/client.nim [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+import chronos, chronicles
+
+import core
+import ../../protocol,
+       ../../../stream/connection,
+       ../../../switch,
+       ../../../utils/future
+
+type
+  DcutrClient* = ref object of RootObj
```
Comment (Menduist) [inline libp2p/protocols/connectivity/dcutr/client.nim:27]:
```suggestion
  DcutrClient* = ref object
```

File: libp2p/protocols/connectivity/dcutr/server.nim [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+      let connectMsg = DcutrMsg.decode(await stream.readLp(1024))
+      debug "Dcutr receiver received a Connect message.", connectMsg
+
+      var ourAddrs = switch.peerStore.getMostObservedProtosAndPorts() # likely empty when the peer is reachable
+      if ourAddrs.len == 0:
+        # this list should be the same as the peer's public addrs when it is reachable
+        ourAddrs =  switch.peerInfo.listenAddrs.mapIt(switch.peerStore.guessDialableAddr(it))
+      var ourDialableAddrs = getTCPAddrs(ourAddrs)
+      if ourDialableAddrs.len == 0:
+        debug "Dcutr receiver has no supported dialable addresses. Aborting Dcutr."
```
Comment (Menduist) [inline libp2p/protocols/connectivity/dcutr/server.nim:47]:
```suggestion
        debug "Dcutr receiver has no supported dialable addresses. Aborting Dcutr.", ourAddrs
```

File: libp2p/protocols/connectivity/dcutr/server.nim [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+      debug "Dcutr receiver has received a Sync message.", syncMsg
+
+      peerDialableAddrs = getTCPAddrs(connectMsg.addrs)
+      if peerDialableAddrs.len == 0:
+        debug "Dcutr initiator has no supported dialable addresses to connect to. Aborting Dcutr."
+        return
+
+      if peerDialableAddrs.len > maxDialableAddrs:
+        peerDialableAddrs = peerDialableAddrs[0..<maxDialableAddrs]
+      var futs = peerDialableAddrs.mapIt(switch.connect(stream.peerId, @[it], forceDial = true, reuseConnection = false))
```
Comment (Menduist) [inline libp2p/protocols/connectivity/dcutr/server.nim:62]: Same as the other side

File: libp2p/protocols/connectivity/dcutr/core.nim (hunk 21 above)
Comment (diegomrsantos) [reply libp2p/protocols/connectivity/dcutr/core.nim:65]: The suggestion is modifying the parameter, isn't it?

File: libp2p/protocols/connectivity/dcutr/core.nim [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+  if r1.isErr or r2.isErr or not checkedEnumAssign(dcutrMsg.msgType, msgTypeOrd):
+    raise newException(DcutrError, "Received malformed message")
+  return dcutrMsg
+
+proc send*(conn: Connection, msgType: MsgType, addrs: seq[MultiAddress]) {.async.} =
+  let pb = DcutrMsg(msgType: msgType, addrs: addrs).encode()
+  await conn.writeLp(pb.buffer)
+
+proc getHolePunchableAddrs*(addrs: seq[MultiAddress]): seq[MultiAddress] =
+  addrs.filterIt(TCP.matchPartial(it))
```
Comment (Menduist) [inline libp2p/protocols/connectivity/dcutr/core.nim:63]: Not sure matchPartial is appropriate, it will also match WS addresses, no?

**Review by Menduist — APPROVED** [review]
LGTM expect last comment to check

**Review by Menduist — APPROVED** [review]

File: libp2p/protocols/connectivity/dcutr/core.nim (hunk 25 above)
Comment (diegomrsantos) [reply libp2p/protocols/connectivity/dcutr/core.nim]: Changed to match.
